### PR TITLE
coverage: Store coverage source regions as `Span` until codegen

### PR DESCRIPTION
**Pull request by Zalathar**
Historically, coverage spans were converted into line/column coordinates during the MIR instrumentation pass.

This PR moves that conversion step into codegen, so that coverage spans spend most of their time stored as `Span` instead.

In addition to being conceptually nicer, this also reduces the size of coverage mappings in MIR, because `Span` is smaller than 4x u32.

---

There should be no changes to coverage output.

### CODE REVIEW

**Comment by jieyouxu** [comment]
Question: is there any kind of rough baseline performance measurements we can compare this against? I understand if that is uhh *non-trivial*.

**Comment by Zalathar** [comment]
> Question: is there any kind of rough baseline performance / profile artifact size measurements we can compare this against?

For coverage instrumentation, we currently have very little of that sort of thing. None of the perf benchmarks enable `-Cinstrument-coverage`, for example.

(Sometimes that's convenient, because it means I don't have to spend time defending coverage-specific regressions caused by necessary refactoring work. But eventually it would be nice to have benchmarks rather than just vibes.)

The main thing we do have is the `coverage-map` tests, which indicate the size of each function's coverage metadata, and also an estimate of how many physical counters it uses (which affects the size of `.profraw` output when running the instrumented binary).

But those metrics are unaffected by this PR, which (presumably) just gives a modest decrease in the amount of memory used to store coverage metadata in MIR.

**Comment by Zalathar** [comment]
All that said, for this particular PR any improvement is just gravy, not an intentional goal.

File: compiler/rustc_codegen_llvm/src/coverageinfo/mapgen/spans.rs [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+        .span_to_source(span, |src, start, end| try {
+            // We're only checking for specific ASCII characters, so we don't
+            // have to worry about multi-byte code points.
+            if try_next && src.as_bytes()[end] == b'{' {
+                Some(span.with_hi(hi + BytePos(1)))
+            } else if try_prev && src.as_bytes()[start - 1] == b'}' {
+                Some(span.with_lo(lo - BytePos(1)))
+            } else {
+                None
+            }
```
Comment (jieyouxu) [inline compiler/rustc_codegen_llvm/src/coverageinfo/mapgen/spans.rs:88]: Remark: `:ferrisAware:` but this is fine.

**Review by jieyouxu — APPROVED** [review]
Thanks, this looks like a nice robustness improvement to me.

**Comment by Zalathar** [comment]
In this case I don't think there's benefit in avoiding rollup:
- None of the benchmarks enable coverage, so a separate perf run wouldn't do anything useful.
- This PR only touches coverage-specific code paths and can't realistically break non-coverage tests.

@bors rollup=maybe

File: compiler/rustc_codegen_llvm/src/coverageinfo/mapgen/spans.rs (hunk 1 above)
Comment (Zalathar) [reply compiler/rustc_codegen_llvm/src/coverageinfo/mapgen/spans.rs:88]: Yeah, I'm very much aware of why `BytePos(1)` is bad news in general; in the past I have fixed `BytePos(1)` bugs in the previous incarnation of this code.

In this very specific set of circumstances, we have confirmed in advance that the code point we're skipping over is a single byte (ASCII `{` or `}`), so adjusting by 1 to skip that character really is OK.

**Pull request merged**